build(docs-infra): introduce new process for generating data for the events page

gkalpak opened the pull request:
This PR introduces a new process for generating data for the AIO [events page](https://angular.io/events), which streamlines the process and minimizes duplication and manual work. For more details, see `aio/scripts/generate-events/README.md`.

**TODECIDE**

- [x] How we want to present dates (esp. the `Starts on` prefix).
    _We will use the `YYYY-MM-DD` format and change the column header to "Start date"_
- [x] What to do about events with invalid dates. (Currently, the script prints them and then removes them from the list.)
    _The current behavior is fine._
- [x] Whether we want to update past events with links (or maybe just the 2022 ones).
    _We can do that, but we also don’t have links for all the events. We’ll try to be mindful for 2022 events from now on. It’s fine if we don’t have links for all the events from the start._
- [x] Whether we want to automatically keep X last years worth of events.
  (We could keep them all in the DB, but not put them in `events.json` for displaying in AIO.)
    _For now, we will display all events (since 2019). In the future, we might want to revisit this and only show a limited number of years or show all available years but as collapsibles._
- [x] How much back we want to go. (NOTE: Sheets before 2019 use different formats).
    _We’re good until 2019._
- [x] Whether to only retrieve events from the DB when building `next.angular.io` on upstream CI and get `next.angular.io/generated.events.json` for other builds.
  (This would allow us to lock down reading on the database fwiw.)
    _We'll go with the current implementation for simplicity. We can revisit in the future if the number of reads becomes problematic._


**TODO**

- [x] Set up Apps Script extension on the actual spreadsheet and create time-driven trigger.
- [x] Ensure the installable trigger creator has access to the `angular-io` Firebase project.
- [x] Update `DB_URL` value in `constants.js` and `index.mjs`.
- [x] Rework some parts of the implementation - see [here][1].
       _PR: angular/dev-infra#738_
- [x] List specific ways to get in touch with DevRel in `aio/content/marketing/events-contributing.md`.
- [x] Link to the GitHub Action that periodically updates events in `aio/scripts/generate-events/README.md`.
- [x] Address TODOs in `.github/workflows/update-events.yml`.
- [x] Update payload sizes in `aio-payloads.json`.

[1]: https://github.com/angular/angular/pull/45588#issuecomment-1145955926

mary-poppins commented:
You can preview 832adb0 at https://pr45588-832adb0.ngbuilds.io/.

mary-poppins commented:
You can preview a551b54 at https://pr45588-a551b54.ngbuilds.io/.

mgechev commented on aio/src/app/custom-elements/events/events.component.html:
```diff
@@ -1,52 +1,47 @@
+          <span *ngSwitchCase="false">{{ event.name }}</span>
+          <a *ngSwitchDefault href="{{ event.linkUrl }}" title="{{ event.tooltip }}">{{ event.name }}</a>
+        </th>
+        <td>{{ event.location ?? '-' }}</td>
```
Maybe we can remove this column? We're not providing a location for now.

mgechev commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -47,56 +47,70 @@ export class EventsComponent implements OnInit {
-      // If no work shop date create conference date string
-      dateString = processDate(event.date);
+      // If no workshops date, create conference date string.
+      const datePrefix = event.date.end ? '' : 'Starts on ';
```
Can we remove the prefix "Starts on"? Maybe we can update the column title to be "Starts on date" to avoid duplication.

mgechev reviewed (changes requested):
Left two comments. Looks really well!

jessicajaniuk reviewed (dismissed):
reviewed-for: size-tracking

mary-poppins commented:
You can preview ed461dc at https://pr45588-ed461dc.ngbuilds.io/.

mary-poppins commented:
You can preview fcce55c at https://pr45588-fcce55c.ngbuilds.io/.

mary-poppins commented:
You can preview 2bb3c74 at https://pr45588-2bb3c74.ngbuilds.io/.

mary-poppins commented:
You can preview 9845308 at https://pr45588-9845308.ngbuilds.io/.

mary-poppins commented:
You can preview 47462ab at https://pr45588-47462ab.ngbuilds.io/.

mary-poppins commented:
You can preview 490d3ba at https://pr45588-490d3ba.ngbuilds.io/.

mary-poppins commented:
You can preview f8f12ab at https://pr45588-f8f12ab.ngbuilds.io/.

mary-poppins commented:
You can preview 3363959 at https://pr45588-3363959.ngbuilds.io/.

mary-poppins commented:
You can preview 28a2e0d at https://pr45588-28a2e0d.ngbuilds.io/.

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

gkalpak commented:
After discussing this with the dev-infra team, we decided to change the implementation as follows in order to keep CI more deterministic and avoid introducing a CI dependency on an external system:

- The Firebase DB will continue to be updated from the spreadsheet as happens currently.
- The `events.json` file will be checked into the repo (and not be generated on every build).
- A GitHub Action will be implemented that will check the Firebase DB daily and (if necessary) send a PR to update `events.json`.

mary-poppins commented:
You can preview 82b95f4 at https://pr45588-82b95f4.ngbuilds.io/.

mary-poppins commented:
You can preview bbb3d2b at https://pr45588-bbb3d2b.ngbuilds.io/.

mary-poppins commented:
You can preview 6b6e3c6 at https://pr45588-6b6e3c6.ngbuilds.io/.

gkalpak committed:
build(docs-infra): introduce new process for generating data for the events page

This commit introduces a new process for generating data for the AIO
[events page](https://angular.io/events), which streamlines the process
and minimizes duplication and manual work. For more details, see
`aio/scripts/generate-events/README.md`.

mary-poppins commented:
You can preview 09fd410 at https://pr45588-09fd410.ngbuilds.io/.

mary-poppins commented:
You can preview c3e545f at https://pr45588-c3e545f.ngbuilds.io/.

mary-poppins commented:
You can preview ab490fc at https://pr45588-ab490fc.ngbuilds.io/.

mary-poppins commented:
You can preview 8969655 at https://pr45588-8969655.ngbuilds.io/.

mary-poppins commented:
You can preview c4cb9f7 at https://pr45588-c4cb9f7.ngbuilds.io/.

mary-poppins commented:
You can preview 8573647 at https://pr45588-8573647.ngbuilds.io/.

gkalpak committed:
fixup! build(docs-infra): introduce new process for generating data for the events page

mary-poppins commented:
You can preview 9ec7d45 at https://pr45588-9ec7d45.ngbuilds.io/.

AndrewKushnir commented:
Merge-assistance: needs a "green light" from @mgechev before landing this PR.

AndrewKushnir commented:
Got a "green light" from @mgechev offline, so we can proceed with merging this PR. // cc @dylhunn

dylhunn commented:
This PR was merged into the repository by commit c9241fdeb0b889f883cb375d0cb822603d233290.

dylhunn commented:
@gkalpak Was this PR related to `update-events.yml`? I have started getting emails that the Update AIO events script has failed to run, with [links like this one](https://github.com/angular/angular/actions/runs/3046168453).

gkalpak commented:
Yes. As discussed "offline", notifications for scheduled workflows go to the last person that modified the schedule: [GitHub docs][1]

Once the errors are fixed, hopefully there won't be any notifications :grin:

[1]: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule:~:text=Notifications%20for%20scheduled%20workflows
